### PR TITLE
*: index codec for common handle

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve?

Problem Summary:

Support index codec for CommonHandle.
Basic code for cluster index.

### What is changed and how it works?

Proposal: [clustered index](https://github.com/pingcap/tidb/pull/17044) 

What's Changed:

Support index codec for CommonHandle.

### Check List 

Tests 

- Unit test

### Release note 
- No release note

### CODE REVIEW

File: table/tables/index.go [hunk 1]
```diff
@@ -261,8 +281,19 @@ func (c *index) GenIndexKey(sc *stmtctx.StatementContext, indexedValues []types.
 //		|
 //		|     Layout: Flag
 //		|     Length:  1
+//		+--Unique Common Handle
```
Comment (wjhuang2016) [inline table/tables/index.go:284]: L271 Without Restore Data(same with old layout)    the comment needs to refine.

File: table/tables/index.go [hunk 2]
```diff
@@ -338,7 +373,16 @@ func (c *index) Create(sctx sessionctx.Context, rm kv.RetrieverMutator, indexedV
 	} else {
 		idxVal = make([]byte, 0)
 		if distinct {
-			idxVal = EncodeHandleInUniqueIndexValue(h.IntValue())
+			if h.IsInt() {
+				idxVal = EncodeHandleInUniqueIndexValue(h.IntValue())
+			} else {
+				if opt.Untouched {
```
Comment (wjhuang2016) [inline table/tables/index.go:379]: Is it duplicate with L387?

File: table/tables/index_test.go [hunk 3]
```diff
@@ -122,7 +130,7 @@ func (s *testIndexSuite) TestIndex(c *C) {
 
 	_, hit, err := index.Seek(sc, txn, values)
 	c.Assert(err, IsNil)
-	c.Assert(hit, IsTrue)
+	c.Assert(hit, IsFalse)
```
Comment (wjhuang2016) [inline table/tables/index_test.go:133]: Why change this?

File: table/tables/index.go (hunk 2 above)
Comment (coocood) [reply table/tables/index.go:379]: No L379 append the tailLen, L387 appends the untouched flag.

File: table/tables/index_test.go (hunk 3 above)
Comment (coocood) [reply table/tables/index_test.go:133]: The original implementation uses 0 as int handle, so the `hit` is true if the handle happens to be 0.
The new implementation doesn't append anything to the index key, so the `hit` is false.

The hit is only used in test code, so it doesn't matter.

File: tablecodec/tablecodec.go [hunk 4]
```diff
@@ -608,67 +611,76 @@ func CutIndexKeyNew(key kv.Key, length int) (values [][]byte, b []byte, err erro
...
+	// HandleNotNeeded means no need to decode handle value when DecodeIndexKV.
+	HandleNotNeeded
 )
 
-func handleExists(hdStatus HandleStatus) bool {
-	return hdStatus != HandleNotExists
-}
-
-// reEncodeHandleByStatus first decode the value into a int or uint decided by the hdStatus, then encode it so that it can
+// reEncodeHandle first decode the value into a int or uint decided by the unsigned, then encode it so that it can
```
Comment (wjhuang2016) [inline tablecodec/tablecodec.go:622]: The comment is outdated.

File: table/tables/index.go (hunk 2 above)
Comment (crazycs520) [reply table/tables/index.go:380]: Please add a comment for this.

Comment (crazycs520) [reply table/tables/index.go:380]: Should move L380 ~ 384 before L377?

File: table/tables/index.go [hunk 5]
```diff
@@ -314,10 +345,15 @@ func (c *index) Create(sctx sessionctx.Context, rm kv.RetrieverMutator, indexedV
...
+		if !h.IsInt() && distinct {
+			idxVal = encodeCommonHandle(idxVal, h)
+		}
+		idxVal = append(idxVal, rowRestoredValue...)
 		tailLen := 0
-		if distinct {
+		if h.IsInt() && distinct {
 			// The len of the idxVal is always >= 10 since len (restoredValue) > 0.
 			tailLen += 8
 			idxVal = append(idxVal, EncodeHandleInUniqueIndexValue(h.IntValue())...)
```
Comment (crazycs520) [inline table/tables/index.go:359]: Should move this line before L354?

File: table/tables/index.go (hunk 2 above)
Comment (coocood) [reply table/tables/index.go:380]: > Should move L380 ~ 384 before L377?

No, if the handle is int, there would be no tailLen at the first byte.

File: table/tables/index.go (hunk 5 above)
Comment (coocood) [reply table/tables/index.go:359]: No.
It's consistent with the comment.
Integer handle is appended at the tail if it has restored value.

**Comment by siddontang** [comment]
image

where is the proposal?

**Comment by coocood** [comment]
> image
> 
> where is the proposal?

It being reviewed https://github.com/pingcap/tidb/pull/17044

**Pull request merged**